src/Avalonia.Vulkan/VulkanDisplay.cs

```diff
@@ -0,0 +1,404 @@
+            CommandBufferPool.Dispose();
+        }
+
+        private static unsafe SwapchainKHR CreateSwapchain(VulkanInstance instance, VulkanDevice device,
+            VulkanPhysicalDevice physicalDevice, VulkanSurface surface, out Extent2D swapchainExtent,
+            VulkanDisplay oldDisplay = null)
+        {
+            if (_swapchainExtension == null)
+            {
+                instance.Api.TryGetDeviceExtension(instance.ApiHandle, device.ApiHandle, out KhrSwapchain extension);
+
+                _swapchainExtension = extension;
+            }
+
+            while (!surface.CanSurfacePresent(physicalDevice))
+            {
+                Thread.Sleep(16);
+            }
+
+            VulkanSurface.SurfaceExtension.GetPhysicalDeviceSurfaceCapabilities(physicalDevice.ApiHandle,
+                surface.ApiHandle, out var capabilities);
+
+            uint presentModesCount;
+
+            VulkanSurface.SurfaceExtension.GetPhysicalDeviceSurfacePresentModes(physicalDevice.ApiHandle,
+                surface.ApiHandle,
+                &presentModesCount, null);
+
+            var presentModes = new PresentModeKHR[presentModesCount];
+
+            fixed (PresentModeKHR* pPresentModes = presentModes)
+            {
+                VulkanSurface.SurfaceExtension.GetPhysicalDeviceSurfacePresentModes(physicalDevice.ApiHandle,
+                    surface.ApiHandle, &presentModesCount, pPresentModes);
+            }
+
+            var imageCount = capabilities.MinImageCount + 1;
+            if (capabilities.MaxImageCount > 0 && imageCount > capabilities.MaxImageCount)
+                imageCount = capabilities.MaxImageCount;
+
+            var surfaceFormat = surface.GetSurfaceFormat(physicalDevice);
+
+            bool supportsIdentityTransform = capabilities.SupportedTransforms.HasFlag(SurfaceTransformFlagsKHR.SurfaceTransformIdentityBitKhr);
+            bool isRotated = capabilities.CurrentTransform.HasFlag(SurfaceTransformFlagsKHR.SurfaceTransformRotate90BitKhr) || 
+                capabilities.CurrentTransform.HasFlag(SurfaceTransformFlagsKHR.SurfaceTransformRotate270BitKhr);
+
+            if (capabilities.CurrentExtent.Width != uint.MaxValue)
+            {
+                swapchainExtent = capabilities.CurrentExtent;
+            }
+            else
+            {
+                var surfaceSize = surface.SurfaceSize;
+
+                var width = Math.Max(capabilities.MinImageExtent.Width,
+                    Math.Min(capabilities.MaxImageExtent.Width, (uint)surfaceSize.Width));
+                var height = Math.Max(capabilities.MinImageExtent.Height,
+                    Math.Min(capabilities.MaxImageExtent.Height, (uint)surfaceSize.Height));
+
+                swapchainExtent = new Extent2D(width, height);
+            }
+
+            PresentModeKHR presentMode;
+
+            if (presentModes.ToList().Contains(PresentModeKHR.PresentModeMailboxKhr))
```

Comment:
I would move this `ToList` out to reuse it later. Could consider `ToHashSet` too but this is a tiny list anyway.

Reply:
Addressed